Apg/sap integration

annapiegra opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "fix bug in ideal gas EOS.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary
Creates a new f90-interface needed for the Pagosa integration.

-Adds eos_setOptions 
-Decouples eospac preinversion and insert data
-Uses a single step instead of two to get sesame P from D+Ut 

TODO: docs, tests, fix properly singularity_eos.f90 spiner deps...

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file
- [ ] If preparing for a new release, update the version in cmake.

jhp-lanl commented:
@annapiegra I fixed a couple errors that were breaking the build. I also added the `MinInternalEnergyFromDensity()` function to the scratch buffer size querrying mechanism. Currently I'm assuming EOSPAC either won't write to the second derivative array (`dFy` in EOSPAC parlance) or that because we don't care about that value, nothing nasty will happen when it does. Feel free to add another scratch buffer if you think that's going to be an issue.

annapiegra commented:
Thanks.  Since it’s a 1d function dFy shouldn’t matter. Also we don’t use the derivatives so we should be good on that front too.I’ll verify later this week. What compiler were you using? I’ve only tested with the intel compiler used by sap.

Sent with BlackBerry Work
(www.blackberry.com)


From: Jeff Peterson ***@***.******@***.***>>
Date: Tuesday, Jul 25, 2023 at 8:41 PM
To: lanl/singularity-eos ***@***.******@***.***>>
Cc: Pietarila Graham, Anna Mataleena ***@***.******@***.***>>, Mention ***@***.******@***.***>>
Subject: [EXTERNAL] Re: [lanl/singularity-eos] Apg/sap integration (PR #278)


@annapiegra<https://urldefense.com/v3/__https://github.com/annapiegra__;!!Bt8fGhp8LhKGRg!FyhuaD2hC3HpePc4nPXZnp4_hKXTt7A0ArtoSCtcAFKj4jw-LI02krJqVh9TevMmKv24Pi7JVIRigWte8djEk7o$> I fixed a couple errors that were breaking the build. I also added the MinInternalEnergyFromDensity() function to the scratch buffer size querrying mechanism. Currently I'm assuming EOSPAC either won't write to the second derivative array (dFy in EOSPAC parlance) or that because we don't care about that value, nothing nasty will happen when it does. Feel free to add another scratch buffer if you think that's going to be an issue.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/lanl/singularity-eos/pull/278*issuecomment-1650887339__;Iw!!Bt8fGhp8LhKGRg!FyhuaD2hC3HpePc4nPXZnp4_hKXTt7A0ArtoSCtcAFKj4jw-LI02krJqVh9TevMmKv24Pi7JVIRigWteZGfjAcs$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AU36G6NESBIYL7RNPHKZZEDXSB7UZANCNFSM6AAAAAA2XHAN7I__;!!Bt8fGhp8LhKGRg!FyhuaD2hC3HpePc4nPXZnp4_hKXTt7A0ArtoSCtcAFKj4jw-LI02krJqVh9TevMmKv24Pi7JVIRigWteq_6fT6w$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Yurlungur commented on eospac-wrapper/eospac_wrapper.cpp:
```diff
@@ -127,19 +129,78 @@ EOS_INTEGER eosSafeLoad(int ntables, int matid, EOS_INTEGER tableType[],
+
+
+  //choice of which table types setOption is called on mimics SAP. Some table types are incmopatible whiel others lead to numerical issues. 
+  if (monotonicity==1 || monotonicity==2) {
```
What is the meaning of these settings? It might be better if we used an enum for monotonicity so we could more easily read what these settings do?

annapiegra replied:
Added enum

Yurlungur commented on eospac-wrapper/eospac_wrapper.cpp:
```diff
@@ -127,19 +129,78 @@ EOS_INTEGER eosSafeLoad(int ntables, int matid, EOS_INTEGER tableType[],
+    }
+  }
+ 
+  if (abs(apply_splitting) == 1) {
```
same as above---maybe an enum would be better here?

annapiegra replied:
added enum

Yurlungur reviewed (approved):
Overall this seems reasonable to me. I'd like Richard's eyes on the EOSPAC stuff and I would also like to make sure all the PR checklist boxes are checked. But barring that, I approve. A few nitpicks below.

annapiegra commented on singularity-eos/eos/eos_gruneisen.hpp:
```diff
@@ -313,6 +314,11 @@ PORTABLE_INLINE_FUNCTION Real Gruneisen::PressureFromDensityInternalEnergy(
 }
+PORTABLE_INLINE_FUNCTION Real Gruneisen::MinInternalEnergyFromDensity(
+    const Real rho_in, Real *lambda) const {
+  const Real rho = std::min(rho_in, _rho_max);
```
Added.

jhp-lanl commented on singularity-eos/eos/eos_base.hpp:
```diff
@@ -262,6 +263,35 @@ class EosBase {
     PressureFromDensityInternalEnergy(rhos, sies, pressures, num,
                                       std::forward<LambdaIndexer>(lambdas));
   }
+  ///
+  template <typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
+  inline void MinInternalEnergyFromDensity(ConstRealIndexer &&rhos, RealIndexer &&sies,
+                                           const int num, LambdaIndexer &&lambdas) const {
+    static auto const name = SG_MEMBER_FUNC_NAME();
+    static auto const cname = name.c_str();
+    CRTP copy = *(static_cast<CRTP const *>(this));
+    portableFor(
+        cname, 0, num, PORTABLE_LAMBDA(const int i) {
+          sies[i] = copy.MinInternalEnergyFromDensity(rhos[i], lambdas[i]);
+        });
+  }
+  template <typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer,
+            typename = std::enable_if_t<!is_raw_pointer<RealIndexer, Real>::value>>
+  inline void MinInternalEnergyFromDensity(ConstRealIndexer &&rhos, RealIndexer &&sies,
+                                           Real * /*scratch*/, const int num,
+                                           LambdaIndexer &&lambdas) const {
+    MinInternalEnergyFromDensity(std::forward<ConstRealIndexer>(rhos),
+                                 std::forward<RealIndexer>(sies), num,
+                                 std::forward<LambdaIndexer>(lambdas));
+  }
+  template <typename LambdaIndexer>
+  inline void MinInternalEnergyFromDensity(const Real *rhos, Real *sies,
+                                           Real * /*scratch*/, const int num,
+                                           LambdaIndexer &&lambdas,
+                                           Transform && = Transform()) const {
+    MinInternalEnergyFromDensity(rhos, num, std::forward<LambdaIndexer>(lambdas));
+  }
+  ///
```
Would it make sense to make a default definition of `MinInternalEnergyFromDensity()` in the EosBase class? It could be something as simple as taking the zero-K isotherm energy at a given density. EOSPAC would of course override this, but I think it would be effective for many of the analytic EOS rather than just returning zero.

annapiegra replied:
Is this something we want implemented asap?

jhp-lanl replied:
I guess since Pagosa only does this for sesame tables, it's not a very big deal right now. But I would _greatly_ prefer that approach to stubbing out the feature for all EOS. It should only be a few lines of code that need to be changed though.

jhp-lanl replied:
Made an issue for this https://github.com/lanl/singularity-eos/issues/318

annapiegra commented:
Formatting check fails because changing the order of hpp files in eospac_wrapper.cpp leads to compiler errors.

jhp-lanl commented:
@cmauney @Yurlungur @dholladay00 @rbberger  I think this is ready to go if any of you have further comments

Yurlungur commented on singularity-eos/eos/modifiers/ramps_eos.hpp:
```diff
@@ -117,7 +117,7 @@ class BilinearRampEOS : public EosBase<BilinearRampEOS<T>> {
   PORTABLE_INLINE_FUNCTION
   Real get_ramp_dpdrho(Real rho) const {
-    const Real dpdr{rho < r0_ ? 0.0 : rho < rmid_ ? ratio(a_, r0_) : ratio(b_, r0_)};
+    const Real dpdr{rho < rmid_ ? ratio(a_, r0_) : ratio(b_, r0_)};
```
Consider removing the branching here if possible by using masks instead?

jhp-lanl replied:
I think we'll defer this since it might make sense to replace these all wholesale in the future.

jhp-lanl replied:
see https://github.com/lanl/singularity-eos/issues/319

Yurlungur reviewed (approved):
Great work all---glad to see this in such good shape. I will let @mauneyc-LANL and @dholladay00 weigh in, but from my perspective, this is ready to go.

jhp-lanl commented:
@cmauney @Yurlungur @dholladay00 @rbberger  can I interpret your silence for approval?

dholladay00 commented on singularity-eos/eos/eos_ideal.hpp:
```diff
@@ -78,6 +78,11 @@ class IdealGas : public EosBase<IdealGas> {
   }
+  PORTABLE_INLINE_FUNCTION Real
+  MinInternalEnergyFromDensity(const Real rho, Real *lambda = nullptr) const {
+    return 0.0;
```
@jhp-lanl could you add the not enabled warning function here?

jhp-lanl replied:
This value is actually valid for the ideal gas. However, an appropriate shift was not implemented in the Shift modifier (which is the only thing that would cause this value to be non-zero). I've added that now

dholladay00 commented on singularity-eos/eos/eos_vinet.hpp:
```diff
@@ -304,6 +306,14 @@ PORTABLE_INLINE_FUNCTION Real Vinet::PressureFromDensityInternalEnergy(
 }
+PORTABLE_INLINE_FUNCTION Real Vinet::MinInternalEnergyFromDensity(const Real rho,
+                                                                  Real *lambda) const {
+#ifndef NDEBUG
```
could probably replace with the function @jhp-lanl

jhp-lanl replied:
done

rbberger commented on spack-repo/packages/singularity-eos/package.py:
```diff
@@ -12,9 +12,9 @@ class SingularityEos(CMakePackage, CudaPackage):
-    homepage = "https://lanl.github.io/singularity-eos/main/index.html"
-    git = "https://github.com/lanl/singularity-eos.git"
-    url = "https://github.com/lanl/singularity-eos/archive/refs/tags/release-1.6.1.tar.gz"
+    homepage = "https://re-git.lanl.gov/xcap/oss/singularity-eos"
```
I don't think these 3 lines should change

jhp-lanl replied:
Fixed

rbberger commented on singularity-eos/eos/eos_eospac.hpp:
```diff
@@ -53,10 +54,91 @@ using namespace eos_base;
+
+inline void SetUpDensityTemperatureScalingOptions(EOS_INTEGER options[],
+                                                  EOS_REAL values[], EOS_INTEGER &nopts,
+                                                  Transform &transform) {
```
This and others can be `const Transform &`

jhp-lanl replied:
Done

rbberger commented on singularity-eos/eos/eos_eospac.hpp:
```diff
@@ -590,72 +638,58 @@ class EOSPAC : public EosBase<EOSPAC> {
+    using namespace EospacWrapper;
+    EOS_REAL *R = const_cast<EOS_REAL *>(&rhos[0]);
+    EOS_REAL *E = &sies[0];
+    EOS_REAL *dedr = scratch + 1 * num;
```
```suggestion
    EOS_REAL *dedr = scratch + 0 * num;
```

rbberger reviewed (approved):
Overall looks good to me. Just some minor things I found.

rbberger commented on singularity-eos/eos/eos_base.hpp:
```diff
@@ -68,6 +68,7 @@ char *StrCat(char *destination, const char *source) {
   using EosBase<EOSDERIVED>::InternalEnergyFromDensityTemperature;                       \
   using EosBase<EOSDERIVED>::PressureFromDensityTemperature;                             \
   using EosBase<EOSDERIVED>::PressureFromDensityInternalEnergy;                          \
+  using EosBase<EOSDERIVED>::MinInternalEnergyFromDensity;                               \
```
Still needs to be added to the Python interface, but that can be its own PR.

jhp-lanl replied:
Issue here: https://github.com/lanl/singularity-eos/issues/321

dholladay00 commented on singularity-eos/eos/eos_eospac.hpp:
```diff
@@ -1331,8 +1323,31 @@ SG_PIF_NOWARN
+  EOS_INTEGER options[]{EOS_Y_CONVERT, EOS_F_CONVERT};
+  EOS_REAL values[]{sieFromSesame(1.0), pressureFromSesame(1.0)};
+  EOS_INTEGER nopts = 2;
+  EOS_REAL R[1] = {rho}, E[1] = {sie}, P[1], dPdr[1], dPde[1], Ec[1];
```
I'm seeing an unused variable warning on `Ec`

jhp-lanl replied:
I'm surprised I didn't get that. What did you need to do to enable that warning?

The unused variable is fixed though

jhp-lanl replied:
Along these lines, should we enable `-Werror` in the CI?

jhp-lanl replied:
@dholladay00 I'm going to go ahead and merge. I fixed the unused variable, but we should discuss whether we should maybe change the CMake to error on warnings and make sure that all appropriate warnings are enabled.

jhp-lanl commented:
Tests pass on gitlab as well as github. Merging.
